### PR TITLE
[fix] [ml] Fix the incorrect total size if use ML interceptor

### PR DESCRIPTION
**Pull request by poorbarcode**
#### 1. Issue

`totalSize` of managed ledger is less than `0`

dump_totalSize_of_ml_less_than_0

----

#### 2. Analizy
Before writing to BK, `OpAddEntry` records the size of the data to `dataLength`, but if the data is changed by `ManagedLedgerInterceptor.processPayloadBeforeLedgerWrite`, the field `dataLength` of `OpAddEntry` is not changed, which will cause the field `totalSize` and `currentLedgerSize` of managed ledger to be incorrect. 

For example:
1. write data `[1]`, before write BK record the data size by `OpAddEntry.dataLength`.
  1-1. now `OpAddEntry.dataLength` is `1`.
2. reset `data` of  OpAddEntry to `[1,2,3...21]` by `ManagedLedgerInterceptor.processPayloadBeforeLedgerWrite`.
  2-1. now the actual size of data is `21`, but `OpAddEntry.dataLength` still is `1`.
3. do write BK and switch the ledger
  3-1. Create ledger info and put it into `ML.ledgers`, the size of ledger info is `21`.
4. trim ledgers
  4-1. `ML.totalSize -= ledgerInfo.size`, then the `totalSize` of ML is `1-21 = -20`

----

#### 3.Reproduce
You can reproduce this issue by test `testTotalSizeCorrectIfHasInterceptor` and `testCurrentLedgerSizeCorrectIfHasInterceptor`

----

#### 4. This issue comes with the new features by PR https://github.com/apache/pulsar/pull/12536

----

### Modifications

Correct variables `OpAddEntry.dataLength` and `ML.currentLedgerSize` after executing `ManagedLedgerInterceptor.processPayloadBeforeLedgerWrite`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 
- https://github.com/poorbarcode/pulsar/pull/62

### CODE REVIEW

**Review by mattisonchao — APPROVED** [review]
LGTM, 
But I think that there are better ways to go than subtracting before adding. Can we move the `currentLedgerSize` calculation logic after the interceptor?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java [hunk 1]
```diff
@@ -138,6 +138,8 @@ public void initiate() {
                     duplicateBuffer = payloadProcessorHandle.getProcessedPayload();
                 }
             }
+            this.dataLength = duplicateBuffer.readableBytes();
+            this.ml.currentLedgerSize += (dataLength - originalDataLen);
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:142]: Is it better to just reset the `dataLength` if the payload processor presents?

```java
if (payloadProcessorHandle != null) {
    duplicateBuffer = payloadProcessorHandle.getProcessedPayload();
    dataLength = duplicateBuffer.readableBytes();
}
```

Comment (poorbarcode) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: Already fixed

**Comment by poorbarcode** [comment]
@mattisonchao 

> Can we move the currentLedgerSize calculation logic after the interceptor?

This will affect `currentLedgerIsFull`'s checking and cause the switching ledger to delay one entry, which I'm not sure is appropriate

**Review by codelipenghui — APPROVED** [review]

**Comment by yaalsn** [comment]
In branch-2.11, it compiles with an error.  @poorbarcode PTAL

```
Error:  COMPILATION ERROR : 
Error:  pulsar/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:[141,32] currentLedgerSize has private access in org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl
```

**Comment by yaalsn** [comment]
In branch-2.11, it causes a compilation failure. @poorbarcode PTAL

`pulsar/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:[141,32] currentLedgerSize has private access in org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl`

**Comment by yaalsn** [comment]
@codelipenghui Please help to review and merge the above two PR otherwise it will affect 2.10 and 2.11 compilation.

**Pull request merged**